### ISSUE TITLE
VM > VM module fails to create UIO device on kernels 5.10 and above.

### ISSUE THREAD

**Issue by JayDoubleu** [gnif/LookingGlass#341]
I'm trying to get the module working on kernels 5.10 (which I require for AMDGPU support for rDNA2 card)
It seems that the /dev/uio device is not getting created on mentioned kernel versions.

So far I have tried:
- Fedora Rawhide (5.10.rc6 kernel) 
- Fedora 33 (5.11.0-0.rc0 kernel)
- Ubuntu 20.04 (5.10 kernel)

In all cases it failed to create the uio device.
Few other observations:

Ubuntu 20.04 with 5.4 kernel and Release/B2 branch - running `make load` works as expected.
Ubuntu 20.04 with 5.4 kernel and Main branch - running `make load` fails to create /dev/uio0
Ubuntu 20.04 with 5.10 kernel and Release/B2 branch + patch to make it compile on 5.10 kernel - `make load` fails to create /dev/uio0 device.

**Comment by gnif** [comment]
If you're on bleeding edge, it's no longer `/dev/uio0` but `/dev/kvmfr0`

**Comment by JayDoubleu** [comment]
Hmm unless I am doing something wrong here, the /dev/kvmfr0 is also not getting created.

Just tried on 
(master branch) - Ubuntu 20.04, kernel 5.10
(master branch) - Fedora 33, kernel 5.11

```
git clone https://github.com/gnif/LookingGlass.git
cd LookingGlass/module/
make
modprobe uio
insmod kvmfr.ko
➜  devel modinfo uio
filename:       /lib/modules/5.11.0-0.rc0.20201217gite994cc240a3b.102.fc33.x86_64/kernel/drivers/uio/uio.ko.xz
license:        GPL v2
rhelversion:    8.99
depends:        
retpoline:      Y
intree:         Y
name:           uio
vermagic:       5.11.0-0.rc0.20201217gite994cc240a3b.102.fc33.x86_64 SMP mod_unload 
sig_id:         PKCS#7
signer:         Fedora kernel signing key
sig_key:        3F:C3:2F:57:9D:BF:E1:8A:3F:59:94:A7:1F:F6:DA:12:9A:62:F9:5B
sig_hashalgo:   sha256
signature:      30:1A:7C:1C:B9:29:28:F5:69:84:A4:AB:39:D2:F0:16:BB:92:96:CD:
		78:5C:14:44:7B:62:10:83:6B:F9:57:48:98:4F:1D:B6:6B:09:32:27:
		D9:C0:4B:81:75:2F:FD:C2:2E:46:04:34:BC:33:5F:E3:0F:89:9B:6E:
		E2:DA:30:E6:73:E1:66:5B:2A:AA:62:6F:11:F2:D6:8D:57:B7:38:41:
		82:50:02:91:16:30:45:61:91:8D:78:3B:85:90:1B:28:85:30:6B:D0:
		BB:11:6B:30:59:6D:12:1A:21:FF:8A:DE:A0:AD:6F:11:BB:4C:85:3B:
		36:62:1F:75:16:53:7D:66:3F:11:0B:12:A8:87:64:59:D3:42:5C:33:
		A4:67:5A:9A:81:1E:90:32:6C:16:DC:C5:83:AD:02:9A:B8:C5:DD:69:
		C8:66:A3:5A:8C:9B:BD:D1:1C:70:9C:2C:DF:EC:D1:CF:8F:77:A0:2E:
		3D:53:C3:69:D2:5D:31:C3:37:51:F7:49:35:C1:38:02:94:42:47:3D:
		71:70:8E:9C:BD:7F:EE:58:4B:ED:0F:2C:2F:AC:37:92:D4:71:D3:CD:
		38:32:15:7B:A0:57:03:6B:01:6A:6B:CD:2F:8F:56:B6:A0:5D:1E:DA:
		28:70:93:E7:CA:66:A4:CE:10:BC:15:A4:06:96:CE:47:6B:2C:36:EC:
		74:BC:13:DC:02:99:4C:5E:8B:92:47:E5:89:0F:93:4C:41:75:C3:D6:
		30:3C:6E:FB:82:30:AB:1C:9A:36:22:E8:EE:CD:83:0E:8F:33:5A:A9:
		DD:A1:5D:FA:AD:9B:EA:A7:62:2E:04:FE:C3:58:94:EB:00:A3:CF:62:
		34:ED:00:8C:F5:7C:B5:B9:50:06:3E:6A:0A:9D:62:F9:D8:1E:F9:87:
		22:F6:EB:93:D9:1B:09:7D:3C:17:B3:70:EE:11:7A:A7:19:AD:BF:11:
		AA:C3:25:EA:18:F4:75:13:A7:7D:56:57:ED:C4:6F:21:08:1E:B2:5A:
		3B:AF:16:62:BE:45:43:A6:B6:50:43:EA:72:C1:BC:48:7F:26:5F:2E:
		90:D6:D1:D2:C4:84:2A:55:3C:99:E0:7E:9E:F5:27:83:B4:8E:D4:7E:
		BB:92:A6:A0:D6:8C:9A:0E:E8:0C:85:C1:55:9F:61:F6:72:01:43:AA:
		84:80:E1:CC:19:22:47:D8:46:60:04:84:5A:82:D3:90:A7:6B:C9:53:
		9C:A0:C4:DF:39:97:1E:15:EC:A4:C1:41:F1:67:CE:49:F2:C9:BB:69:
		D9:03:DF:B1:BE:78:E2:6F:78:38:7D:55:D0:FF:AB:03:1D:6C:23:3C:
		DE:F0:F4:1B:3A:D5:F7:EF:1A:D1:C5:87

➜  modinfo kvmfr
filename:       /lib/modules/5.11.0-0.rc0.20201217gite994cc240a3b.102.fc33.x86_64/extra/kvmfr.ko.xz
author:         Geoffrey McRae <geoff@hostfission.com>
license:        GPL v2
rhelversion:    8.99
alias:          pci:v00001AF4d00001110sv*sd*bc*sc*i*
depends:        
retpoline:      Y
name:           kvmfr
vermagic:       5.11.0-0.rc0.20201217gite994cc240a3b.102.fc33.x86_64 SMP mod_unload 

➜  ls /dev/kvm*
/dev/kvm


```

**Comment by gnif** [comment]
Are you using VM to VM? or VM to Host? The module is only for VM to VM at this time.

**Comment by JayDoubleu** [comment]
VM to VM

**Comment by gnif** [comment]
Does your Linux guest have a IVSHMEM device?

**Comment by JayDoubleu** [comment]
Yes.
`00:01.0 RAM memory: Red Hat, Inc. Inter-VM shared memory (rev 01)`

**Comment by gnif** [comment]
Please show the output of:
```lspci -vv -s 00:01.0```

**Comment by JayDoubleu** [comment]
```
00:01.0 RAM memory: Red Hat, Inc. Inter-VM shared memory (rev 01)
	Subsystem: Red Hat, Inc. QEMU Virtual Machine
	Control: I/O+ Mem+ BusMaster- SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx-
	Status: Cap- 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
	Region 0: Memory at c144b000 (32-bit, non-prefetchable) [size=256]
	Region 2: Memory at 1012000000 (64-bit, prefetchable) [size=32M]
	Kernel modules: kvmfr
```

**Comment by gnif** [comment]
`Kernel modules: kvmfr`

It's not loaded, you should also see:

`Kernel driver in use: kvmfr`

**Comment by JayDoubleu** [comment]
```
➜  module git:(master) ✗ lspci -nn -s 00:01.0


00:01.0 RAM memory [0500]: Red Hat, Inc. Inter-VM shared memory [1af4:1110] (rev 01)
```

**Comment by gnif** [comment]
I see no reason why the `kvmfr` module would not claim the device, please check the usual such as blacklisting, etc. Ensure your VM doesn't have secure boot enabled also as it may prevent this unsigned module from loading.

**Comment by JayDoubleu** [comment]
hmm, both VMs I'm trying it on don't have secure boot enabled and both are vanilla installs.

Only blacklists I have are the ones that come with the system like so:
```
blacklist appletalk
blacklist atm
blacklist ax25
blacklist batman-adv
blacklist floppy
blacklist l2tp_eth
blacklist l2tp_ip
blacklist l2tp_netlink
blacklist l2tp_ppp
blacklist netrom
blacklist nfc
blacklist rds
blacklist rose
blacklist sctp
```

Also, the Release/B2 works on kernel 5.4.0 on the same machine I am testing it.
I even tried the Release/B2 on 5.10 by adding these few extra lines from your commit which allows it to compile and that still does not bring up the device.

Any other places/things I could check for the cause?

**Comment by gnif** [comment]
The fact that `lspci` doesn't even show the module has claimed the device indicates a more fundamental issue with your system, this is very basic kernel module loading boilerplate code that every module uses.

**Comment by JayDoubleu** [comment]
Ok few things.
Does this only require the ivshmem device to be passed through yeah ?
If so if I:
- Create brand new libvirt VM on a separate host
- Attach ivshmem device as per instructions
- Build the module

Should this be sufficient to bring up the /dev/kvmfr0

I'm struggling to see how it's a "fundamental issue with my system" when it works on B2 and 5.4 kernel.
If you confirm that the above is a sufficient test I'm happy to test it on separate machine.

**Comment by gnif** [comment]
> ```
> 00:01.0 RAM memory: Red Hat, Inc. Inter-VM shared memory (rev 01)
> 	Subsystem: Red Hat, Inc. QEMU Virtual Machine
> 	Control: I/O+ Mem+ BusMaster- SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx-
> 	Status: Cap- 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
> 	Region 0: Memory at c144b000 (32-bit, non-prefetchable) [size=256]
> 	Region 2: Memory at 1012000000 (64-bit, prefetchable) [size=32M]
> 	Kernel modules: kvmfr
> ```

The module did not claim the device. I am assuming you `modprobe`d it and checked for it in `lsmod` after?

**Comment by JayDoubleu** [comment]
Well I tried modprobe, I've also tried adding it to /etc/modules so it does it at boot
I also tried insmod directly

**Comment by gnif** [comment]
Did you check `lsmod` to see if it's loaded?

**Comment by JayDoubleu** [comment]
yep. lsmod and modinfo. 
Tried at boot, tried rmmod and reload each and every way.

**Comment by gnif** [comment]
`modinfo` doesn't tell you if it's loaded, just the details about the module.
`lsmod` will show the loaded modules.

Please show the patch you're using to make it load for 5.10?

**Comment by JayDoubleu** [comment]
https://github.com/gnif/LookingGlass/commit/69b8c4b4eb9d36459062cdaaaf88f32366fc3ef0#diff-abb197b3a1f4f4376b4c3971520a1ded57d94c6216c903e7e45da94f4665c9d0R280-R287

Just added the if else bit and built on 5.4

**Comment by gnif** [comment]
bleeding edge doesn't use or even implement `uio` at all anymore. Can you please confirm that the bleeding edge version doesn't claim the device?

**Comment by JayDoubleu** [comment]
if bleeding edge means master branch then yes. I tried it first, multiple times. I only tried B2 to see if this is kernel version issue or not.

**Comment by gnif** [comment]
but were you checking for /dev/uio0 or /dev/kvmfr0?

**Comment by JayDoubleu** [comment]
well, yesterday I spend ours checking for uio0, now I am checking for kvmfr0 and still the same ;)

**Comment by gnif** [comment]
ok, I will make some time to build a 5.10 kernel later and try this.

**Comment by JayDoubleu** [comment]
Thanks, I used this for testing http://ubuntuhandbook.org/index.php/2020/12/install-linux-kernel-5-10-ubuntu-linux-mint/

**Comment by JayDoubleu** [comment]
Just as an observation ( I have no idea what I'm doing)
On 5.4 kernel and Release/B2:

```
       kdev->pgmap.res.flags = pci_resource_flags(dev, 2);
       kdev->pgmap.type      = MEMORY_DEVICE_DEVDAX;
```
Removing these two lines > same behaviour as on 5.10
Bring then back on > works  `Kernel driver in use: kvmfr`

**Comment by gnif** [comment]
Fixed in 13138588896de6a495c341950e3f06be33bd728c
